### PR TITLE
perf: root self-loop skip fast path

### PR DESCRIPTION
**Pull request by ahrav**
Stacked on #4. The single largest win in the series.

## Insight
On real text the automaton stays in the **root** state for the vast majority of bytes (~93% in the Ibsen benchmark), and the root state never emits a match. Those bytes can be skipped in bulk with no dictionary checks.

## Change
`buildRootSkip` derives, from the transition table, the set of bytes that leave the root state, and picks a skip strategy:
- **Single stop byte** — an inlined SWAR word scan handles short gaps; long gaps escalate to the vectorized `bytes.IndexByte`. (Portable analog of an x86-only SIMD prefilter — this target is arm64.)
- **Zero or several stop bytes** — an 8-wide OR over a 0/1 `rootStop` lookup table: one branch per eight input bytes.

`Walk` is specialized into `walkStopByte` / `walkTable` and dispatches on the stop-byte count. `Match` still runs through `Walk`, so it inherits the skip unchanged.

## Verification
- `go test ./...`, `go test -race -run TestDifferentialRandom ./...` — pass
- `BenchmarkMatchIbsen` vs #4 (0 allocs/op throughout):

  | Input | #4 | this PR | speedup |
  |---|---|---|---|
  | 100 B | 323 ns | 111 ns | 2.9x |
  | 1 KB | 3082 ns | 678 ns | 4.5x |
  | 10 KB | 30.7 µs | 7.50 µs | 4.1x |
  | 100 KB | 319 µs | 128 µs | 2.5x |

## Summary by CodeRabbit

* **New Features**
  * Enhanced Aho–Corasick matching with an adaptive “root skip” optimization to speed up scans in the root state.
  * Added bounded trie decoding via `DecodeWithMaxStates` and `DecodeMaxStates`.
* **Bug Fixes**
  * Decoder now performs stronger validation of stream metadata and rejects corrupt/truncated inputs with clearer “corrupt trie” errors.
  * Root-skip initialization is applied consistently when building/decoding tries.
* **Tests**
  * Added extensive correctness tests covering skip behavior, edge cases, and split boundaries.
* **Chores**
  * Added root-skip benchmarks with real-corpus and density/cardinality sweeps.

### CODE REVIEW

File: stream.go [hunk 1]
```diff
@@ -127,11 +127,13 @@ func (dec *decoder) decode() (*Trie, error) {
...
-	return &Trie{
+	trie := &Trie{
 		failTrans: failTrans,
 		dictLink:  dictLink,
 		dict:      dict,
 		pattern:   pattern,
 		bufPool:   newBufPool(),
-	}, nil
+	}
+	trie.buildRootSkip()
```
[the remark this answers is not in the document]
Comment (ahrav) [reply stream.go:186]: Fixed in `0b0115a`. `Decode` now validates the length prefixes before allocating/constructing: it requires all four per-state arrays (`dict`/`failTrans`/`dictLink`/`pattern`) to share one length and to have at least two states (the unused state 0 plus the root), and returns

```
ahocorasick: corrupt trie: inconsistent table lengths (dict=.. failTrans=.. dictLink=.. pattern=..)
```

instead of reaching `buildRootSkip` and panicking on `failTrans[rootState]`. Verified: a crafted stream declaring `failTransLen=1` now returns that error rather than an index-out-of-range panic, and the normal encode/decode round-trip still passes.

(Scoped to length/row-count validation, which is what makes `Decode` total here — `buildRootSkip` on this branch only needs the root row to exist. Per-transition target bounds-checking isn't added because on this branch the wire holds raw state ids with no flag bits to mask, and buildRootSkip doesn't dereference targets.)

File: stream.go [hunk 2]
```diff
@@ -101,6 +102,14 @@ func (dec *decoder) decode() (*Trie, error) {
 		return nil, err
 	}
 
+	// Decode operates on untrusted input. A well-formed trie has one row per
+	// state across all four arrays and at least the unused state 0 plus the
+	// root, so buildRootSkip can index failTrans[rootState]. Reject anything
+	// else with an error rather than panicking on a truncated or corrupt stream.
+	if failTransLen < 2 || dictLen != failTransLen || dictLinkLen != failTransLen || patternLen != failTransLen {
+		return nil, fmt.Errorf("ahocorasick: corrupt trie: inconsistent table lengths (dict=%d failTrans=%d dictLink=%d pattern=%d)", dictLen, failTransLen, dictLinkLen, patternLen)
+	}
```
[the remark this answers is not in the document]
Comment (ahrav) [reply stream.go:141]: Fixed in `981a53b`. Added an up-front `maxDecodeStates = 1 << 24` bound: `Decode` now returns

```
ahocorasick: corrupt trie: N states exceeds decode limit 16777216
```

before any `make(...)`, so an internally consistent but huge count can no longer reach the allocation and panic on slice-allocation limits.

On the value: I measured the repo's own largest automaton — the full `NSF-ordlisten.cleaned.uniq.txt` list (593,526 patterns) builds to **1,173,947 states** — so `1<<24` (16.7M) keeps ~14x headroom and never rejects a realistic trie, while capping the crafted worst case. Verified a stream declaring `1<<25` states now returns the error instead of panicking, and the existing encode/decode round-trip (`TestReadAndWriteTrie`, 10k patterns → 20,402 states) still passes.

Note this bounds the deterministic allocation-limit *panic*; fully eliminating a large-but-sub-cap transient allocation on adversarial input would require streaming the payload instead of length-prefixed allocation, which is a larger change I've left out of scope.

File: rootskip_bench_test.go [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+			pats = append(pats, ws...)
+			got++
+		}
+		if got < card {
+			continue // not enough distinct first letters available
+		}
+		trie := NewTrieBuilder().AddStrings(pats).Build()
+		b.Run(itoaCard(card), func(b *testing.B) { benchMatchOver(b, trie, hay) })
+	}
+}
```
[the remark this answers is not in the document]
Comment (ahrav) [reply rootskip_bench_test.go]: Fixed in `2c363a4`. `BenchmarkRootSkip_CardinalitySweep` now computes the actual stop-byte density each pattern set induces on the fixed Ibsen haystack — a single `O(len(hay))` pass over a `[256]bool` first-byte set — and folds it into each subtest label. The confound is now visible directly in the output:

```
firstbytes_1/density_3pct
firstbytes_2/density_4pct
firstbytes_5/density_18pct
firstbytes_13/density_35pct
firstbytes_26/density_59pct
```

so the cardinality and density effects can be told apart when reading the sweep (and it confirms the confound is real — density rises steeply with cardinality). The header comment no longer claims density is held fixed; it now states the opposite and points at the reported per-subtest density.

File: trie.go [hunk 4]
```diff
@@ -81,25 +96,194 @@ func newBufPool() sync.Pool {
...
+	dict := tr.dict
+	pattern := tr.pattern
+	dictLink := tr.dictLink
+
+	s := rootState
 
+	// The root self-loop skip only pays off when self-looping runs are
+	// long, i.e. at low stop-byte density; at high density it regresses.
+	// Probe a bounded prefix once and drop the skip when it would not help.
+	skip := tr.rootSkipWorthwhile(input)
```
[the remark this answers is not in the document]
Comment (ahrav) [reply trie.go]: Already addressed on this branch in `774bdcd`, before this review round. The eager up-front `rootSkipWorthwhile` prefix scan is gone; `walkTable` (`trie.go:247-274`) now measures stop-byte density **inline** as it skips. It begins optimistic (`skip = true`) and only accumulates the density counters over bytes the skip already reads, disabling the skip once the `rootSkipSampleLen` budget is spent *and* density is above the ~1/16 break-even. So an early-terminating caller — `Walk` whose callback returns `false`, which is how `MatchFirst` stops — with a match near the start returns immediately; nothing is scanned before the first callback fires. That is the lazy/adaptive behavior you suggested.

File: rootskip_bench_test.go [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+				return true
+			}
+		}
+		return false
+	}
+	var pats []string
+	for _, w := range all {
+		if inStop(w[0]) {
+			pats = append(pats, w)
+			if len(pats) == 10000 {
```
[the remark this answers is not in the document]
Comment (ahrav) [reply rootskip_bench_test.go]: Already addressed on this branch in `774bdcd`, before this review round. `BenchmarkRootSkip_DensitySweep_MultiByte` (`rootskip_bench_test.go:167-195`) no longer collects the first 10000 words matching `a..e` (which, on the sorted list, would all be `a` words and yield one stop byte). It buckets words by first byte and takes an equal quota (`per := 10000 / len(stop)`) from each of `a, b, c, d, e`, so the built trie has five root stop bytes and `Match` takes the `walkTable` OR-table path as intended.

File: rootskip_test.go [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+	for _, sz := range []int{1100, 2048, 20000, 50000, 131072} {
+		in := bytesFill(sz, ' ')
+		plant := func(pos int, s string) {
+			if pos >= 0 && pos+len(s) <= len(in) {
+				copy(in[pos:], s)
+			}
+		}
+		mid := sz / 2
+		for d := -maxLen; d <= maxLen; d++ {
+			plant(mid+d, "needle")
```
[the remark this answers is not in the document]
Comment (ahrav) [reply rootskip_test.go]: Already addressed on this branch in `774bdcd`, before this review round. `TestRootSkipSplitBoundaries` (`rootskip_test.go:199-238`) now uses a `planted(sz, pos, s)` helper that returns a **fresh** self-loop haystack with a single pattern copied in. Every midpoint offset (`mid+d`) and every `{abcdefghij at cb-3, xyz at cb}` chunk-boundary case is built in its own buffer, so no plant overwrites another and each subtest runs `checkAgainstNaive` on exactly the boundary case it claims to cover.

File: rootskip_test.go [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+			checkAgainstNaive(t, []string{"cd", "ef"}, in2)
+		})
+	}
+}
+
+// TestRootSkipSplitBoundaries plants matches exactly at and around the input
+// midpoint and every 1/8 chunk boundary, over inputs large enough to trigger
+// the dual-cursor scan (>1KB) and the parallel scan (>16KB). On the current
+// branch these route through the sequential path; on the parallel/dual-cursor
+// PRs later in the stack this is the test that catches an off-by-one in the
```
[the remark this answers is not in the document]
Comment (ahrav) [reply rootskip_test.go]: Fixed in `2c363a4`. Reworded the `TestRootSkipSplitBoundaries` doc comment to describe the durable invariant the test protects — a match that straddles a chunk boundary must be reported exactly once, neither dropped nor duplicated, however the input is partitioned — instead of referencing this branch and the later stack PRs. The size thresholds are kept only as neutral rationale for why the inputs span >1KB and >16KB.

File: stream.go [hunk 8]
```diff
@@ -101,6 +102,24 @@ func (dec *decoder) decode() (*Trie, error) {
...
+	// state across all four arrays and at least the unused state 0 plus the
+	// root, so buildRootSkip can index failTrans[rootState]. Reject anything
+	// else with an error rather than panicking on a truncated or corrupt stream.
+	//
+	// maxDecodeStates also bounds the up-front allocation: failTrans costs 1KB
+	// per state, so an internally consistent but huge count would otherwise
+	// reach make(...) and panic on slice-allocation limits before any data is
+	// read. The bound is far above any realistic automaton (the full NSF word
+	// list in test_data is ~1.2M states), so it never rejects a real trie.
+	const maxDecodeStates = 1 << 24
```
[the remark this answers is not in the document]
Comment (ahrav) [reply stream.go]: Fixed in `bde93ce`. Two changes on the decode path:

1. **Byte-budgeted cap.** The limit is now expressed as a 4 GiB budget for `failTrans` rows and derived back into a state count (`maxDecodeStates = 4<<30 / (256*4)` ≈ 4.19M states), replacing the old `1 << 24` (which was ~16 GiB of rows). The guard now reflects the real allocation size.
2. **Rows read straight into place.** The transient `flatFailTrans := make([]uint32, failTransLen*256)` intermediate (which doubled peak allocation to ~32 GiB at the cap) is gone — `failTrans` rows are decoded one at a time directly via `binary.Read(r, ..., failTrans[i][:])`, so peak allocation equals the budget.

Net: a corrupt-but-consistent length can no longer drive an up-front ~16-32 GiB allocation before any transition payload is read. The 4 GiB budget is ~3.5× the largest real automaton in `test_data` (the full NSF word list, ~1.2M states ≈ 1.2 GiB of rows), so it never rejects a legitimate trie.

File: rootskip_test.go [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+	}
+	// Sprinkle full patterns so there is real match work too.
+	for pos := 0; pos+6 < n; pos += 137 {
+		copy(high[pos:], "banana")
+	}
+
+	t.Run("low_density_skip_engaged", func(t *testing.T) {
+		checkAgainstNaive(t, patterns, low)
+	})
+	t.Run("high_density_skip_disabled", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ahrav) [reply rootskip_test.go]: Good catch — you're right on both counts: the gate is output-invisible, and the cycling `a..e` input keeps the automaton in root-*child* states after the first byte, so root-state sampling may never cross the disable threshold. As written the test could pass without `walkTable` ever disabling the skip. Fixed in `bde93ce`:

- Extracted the gate decision into a behavior-preserving `rootSkipSampler` (`trie.go`), the exact sampler `walkTable` now drives, so the decision has a white-box seam. Added `TestRootSkipSamplerDisableDecision` (`rootskip_test.go`), which asserts the skip is disabled at high density (gap=1 runs, 1/2 density) and left enabled at low density (gap=1024 runs, ~1/1025) once the sample budget is spent — the disable decision is now verified directly rather than through output.
- Reworded `TestRootSkipDensityGate` to an honest correctness-across-densities check (subtests `low_density`/`high_density`): it only asserts `Match` equals the naive reference at either density and no longer claims to verify the (unobservable) disable.

The refactor is arithmetic-identical to the previous inline gate; the differential test and full suite pass.

File: rootskip_test.go [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+	low := rootSkipSampler{budget: rootSkipSampleLen}
+	for i := 0; i < 100; i++ {
+		if low.observe(1024) {
+			t.Fatalf("low density: gate disabled the skip after run %d", i)
+		}
+	}
+	if low.disabled {
+		t.Fatal("low density: gate should leave the skip enabled")
+	}
+}
```
[the remark this answers is not in the document]
Comment (ahrav) [reply rootskip_test.go:339]: Fixed in `e8bb4d6` — added boundary-adjacent cases straddling the exact flip point. One correction on the direction, though: the threshold `stopBytes*16 >= rootBytes` for a constant gap `g` reduces to `16 >= g+1`, i.e. it disables when `g <= 15`. So the flip is between **gap=15** (density 1/16, `16 >= 16` → disables) and **gap=16** (density 1/17, `16 >= 17` false → stays enabled).

The suggested diff had these inverted — `gap=14` (density 1/15, *above* the 1/16 break-even) actually disables, and `gap=16` (below it) stays enabled. I used gap=15 (asserts disable) and gap=16 (asserts enabled), which sit exactly on either side of the `stopBytes*16 >= rootBytes` boundary, so a regression in that arithmetic is now caught. Verified: both new cases pass.

File: stream.go [hunk 11]
```diff
@@ -101,20 +102,45 @@ func (dec *decoder) decode() (*Trie, error) {
...
+	// state, read straight into place below (no transient flat copy), so the
+	// state count is also the peak allocation. Bound it by an explicit byte
+	// budget so the guard matches the real allocation size and a
+	// corrupt-but-consistent length cannot OOM the process before any
+	// transition data is read. The budget is far above any realistic automaton
+	// (the full NSF word list in test_data is ~1.2M states ≈ 1.2 GiB of rows),
+	// so it never rejects a real trie.
+	const (
+		failTransRowBytes = 256 * 4 // one [256]uint32 transition row
+		maxDecodeBytes    = 4 << 30 // 4 GiB budget for failTrans rows
```
[the remark this answers is not in the document]
Comment (ahrav) [reply stream.go]: Fixed in `3c13f93`. You're right that a fixed budget plus an up-front `make([][256]uint32, failTransLen)` still lets a tiny stream reserve the whole ceiling before the read fails. Two changes:

1. **Incremental `failTrans` allocation.** Rows are now appended as they are read (from a bounded ~1 MiB initial capacity), instead of allocating the full declared count up front. A stream that declares `failTransLen == maxStates` but carries little data now only reserves memory proportional to the rows actually delivered, then hits EOF and returns an error — it can no longer force a multi-GiB zeroed allocation before `binary.Read` fails. This directly removes the amplification you described.

2. **Caller-configurable limit.** Added `DecodeWithMaxStates(r, maxStates)` and the exported default `DecodeMaxStates`; `Decode` delegates with the default. Since the library can't know a given process's memory budget, a caller deserializing untrusted input on a memory-constrained process can now bound the decode to what it can afford rather than being stuck with the library's fixed ceiling. A non-positive `maxStates` falls back to the default.

Tests added: `TestDecodeRejectsOversizedStateCount` (count above the limit rejected before allocation), `TestDecodeWithMaxStatesEnforcesCallerLimit` (tight caller limit rejects, default accepts), and `TestDecodeTruncatedBeforeFailTrans` (truncated stream errors via the incremental path). All pass; existing round-trip tests unchanged.
